pkg/cmd/drtprod/cli/commands/yamlprocessor.go

```diff
@@ -137,50 +140,97 @@ func setEnv(environment map[string]string, displayOnly bool) error {
+			defer waitGroupTracker[theTarget.TargetName].Done()
+			defer wg.Done()
+			for _, dt := range theTarget.DependentTargets {
+				if twg, ok := waitGroupTracker[dt]; ok {
+					fmt.Printf("%s: waiting on <%s>\n", theTarget.TargetName, dt)
+					// wait on the dependent targets
+					// it would not matter if we wait sequentially as all dependent targets need to complete
+					twg.Wait()
+				}
+			}
+			err := executeCommands(ctx, theTarget.TargetName, theTarget.commands)
 			if err != nil {
-				fmt.Printf("%s: Error executing commands: %v\n", tn, err)
+				fmt.Printf("%s: Error executing commands: %v\n", theTarget.TargetName, err)
 			}
-			wg.Done()
-		}(targetName, cmds)
+		}(t)
 	}
+	// final wait for all targets to complete
 	wg.Wait()
 	return nil
 }
 
+// buildTargetCmdsAndRegisterWaitGroups builds the commands per target and registers the target to a wait group
+// tracker and returns the same.
+// The wait group tracker is a map of target name to a wait group. A delta is added to the wait group that is
+// marked done when the specific target is complete. The wait group is use by the dependent targets to wait for
+// the completion of the target.
+func buildTargetCmdsAndRegisterWaitGroups(
+	targets []target, targetNameMap map[string]struct{}, targetNames []string,
+) (map[string]*sync.WaitGroup, error) {
+	// map of target name to a wait group. The wait group is used by dependent target to wait for the target to complete
+	waitGroupTracker := make(map[string]*sync.WaitGroup)
+
+	// iterate over all the targets and create all the commands that should be executed for the target
+	for i := 0; i < len(targets); i++ {
+		// expand the environment variables
+		targets[i].TargetName = os.ExpandEnv(targets[i].TargetName)
+		t := targets[i]
+		for j := 0; j < len(t.DependentTargets); j++ {
+			targets[i].DependentTargets[j] = os.ExpandEnv(targets[i].DependentTargets[j])
+		}
+		if _, ok := targetNameMap[t.TargetName]; len(targetNames) > 0 && !ok {
+			fmt.Printf("Ignoring execution for target %s\n", t.TargetName)
+			continue
+		}
+		// add a delta wait for this target. This is added here so that when the execution loop is run, we need not
+		// worry about the sequence
+		waitGroupTracker[t.TargetName] = &sync.WaitGroup{}
+		waitGroupTracker[t.TargetName].Add(1)
+		// Generate the commands for each target's steps
+		targetSteps, err := generateCmdsFromSteps(t.TargetName, t.Steps)
+		if err != nil {
+			return waitGroupTracker, err
+		}
+		targets[i].commands = targetSteps
+	}
+	return waitGroupTracker, nil
+}
+
 // displayCommands prints the commands in stdout
-func displayCommands(name string, cmds []*command) {
-	fmt.Printf("For target <%s>:\n", name)
-	for _, cmd := range cmds {
+func displayCommands(t target) {
+	dependentTargetMsg := ""
+	if len(t.DependentTargets) > 0 {
+		dependentTargetMsg = fmt.Sprintf(" after [%s]", strings.Join(t.DependentTargets, ", "))
+	}
+	fmt.Printf("For target <%s>%s:\n", t.TargetName, dependentTargetMsg)
```

Comment:
Maybe
```
  if len(t.DependentTargets) > 0 {
		fmt.Printf("For target <%s> after [%s]:\n", t.TargetName, strings.Join(t.DependentTargets, ", "))
	} else {
		fmt.Printf("For target <%s>:\n", t.TargetName)
	}
```

Reply:
I was avoiding the repeat. I can put it this way as well.